rango-exchange/rango-client
# chore: upgrade publish scripts

# Summary

We had some troubles on publishing production, the main reason was it hadn't proper error handling and rollback strategies. In this PR I addressed this problems and tried to make scripts more readable and maintainable by restructuring them and add docs. 

Fixes RF-1023

# Changelist

- All the functions related to a command (e.g. `publish`) was inside a file called `utils.mjs` besides the `command.mjs` file. But now they functions related to a specific tool (git,npm,...) has a separate file in `/common` directory. the command itself has been broken into smaller steps and has a separate file besides the `command.mjs`.
- For libs, `esbuild` supports for build the libs in parallel. I'm building the libs in parallel right now and it speed up the process and reduced the workflow time -50% to 70%.
- I fairly added some good comments on `scripts/publish/command.mjs` and hope it communicate the steps wll.
- The current publish has feature management as well, we can enable/disable features on `scripts/common/features.mjs` easily without modifying code.
- All the errors has been defined in `scripts/common/errors.mjs`. it helps us more recognizable errors and also doing proper action when something is failing.
- Most of the functions has a JSDoc comment.
- I added a `State` which simply check the current state of packages to be publish (version, git tag, ...) and then calculate the next state by increasing version. at the end we will compare current state and next state to find problems sooner and bail out the process at the first and don't let the process to start.
- I changed `release` script name to `release-prod` to clearly communicate it should be run on prod only and also avoiding name confilicting with yarn's `release` lifecycle.
- I also added `post-release-prod` which should be run after the green pipeline to sync the latest prod version to our next branch.

# How did you test this change?

You should fork the repository. And rename all `@rango-dev/...` to `@your-org/...`. (You can use VSCode find and replace feature)

Requirements:
- Create an NPM account and an organization. (I created @yeager-dev for example)
- Get a token from NPM and set it as `NPM_TOKEN` in our repo.  
- Enable Github Actions on your fork.
- `post-release-prod` needs add your org name to `scripts/common/constants.mjs` 

You can also test it locally by putting your npm token to `~/.npmrc`.  

Add this line:
```
//registry.npmjs.org/:_authToken=YOUR_NPM_TOKEN
```

# Checklist:
